### DISCUSSION TITLE
Could hb_shape() shape Characters represented by GlyphID

### DISCUSSION THREAD

**Discussion by totake0224** [harfbuzz/harfbuzz#4170] [Q&A] [answered]
I have alternative characters got by hb_ot_layout_lookup_get_glyph_alternates().
and I want shape Text contains one of them.Could I do such a thing with Harfbuzz?
For example, 
```
there is a Text "Abc" and character 'A' has alternatives like 'A`'.
 If I want shape Text "A`bc". but hb_ot_layout_lookup_get_glyph_alternates() return Glyph ID.
and there seems no function to get unicode from Glyph ID. and there seems no hb_buffer_add for Glyph ID.
```
Should I make table to convert Glyph ID to unicode?

If someone knows about this.Please tell me.

Thank you.

**Comment by totake0224** [comment, marked as answer]
I found 
https://github.com/harfbuzz/uharfbuzz/issues/22
I will check this.

**Reply by totake0224** [reply]
This link is perfect. thank you great predecessor.

**Comment by kellyo-cricut** [comment]
Is there a way to do this without the `uharfbuzz` wrapper? I'm able to get an alternate glyph shaped by providing `hb_shape()` a feature, but it seems like I need to determine which feature the glyph id is included in and potentially the index it appears in that feature. I have this sort of working by including the `aalt` feature when shaping, but I don't know how to determine the `value` to set in that feature so that the correct glyph id is substituted. 

My use case is similar to the original question- I know the glyph id that I want to be substituted when shaping, but don't know how to include that in the buffer passed to `hb_shape()` or how to select the correct feature to specify in that call. Any suggestions?

**Reply by khaledhosny** [reply]
You can do the same trick in C, but in general this is something you shouldn’t be doing.

**Reply by kellyo-cricut** [reply]
Is there a better approach for that use case? I have rtf data that specifies an alternate glyph id for a range, absent of any OT feature information.

**Reply by khaledhosny** [reply]
Interesting, how does that even work good ven that glyph ids are specific to a given version of a given font. Any way, you probably don't need to use hb_shape() at all and should just output the glyph as is and use hb_font_t APIs to get its advance.

**Reply by behdad** [reply]
I suppose the font is embedded. And it might be that glyph indices are interspersed with text to be shaped.

Both approaches should work.

I think SVG text is similar.

**Reply by kellyo-cricut** [reply]
Oh yeah, that is correct. The font containing those glyph ids is part of the data that I have, not just random glyph ids by themselves.

I was able to make this work with that C code, so thank you! ❤️
